src/osd/PrimaryLogPG.cc

```diff
@@ -8761,15 +8761,13 @@ void PrimaryLogPG::do_osd_op_effects(OpContext *ctx, const ConnectionRef& conn)
   ceph_assert(conn);
 
   auto session = conn->get_priv();
-  if (!session)
-    return;
 
   for (list<pair<watch_info_t,bool> >::iterator i = ctx->watch_connects.begin();
        i != ctx->watch_connects.end();
        ++i) {
     pair<uint64_t, entity_name_t> watcher(i->first.cookie, entity);
     dout(15) << "do_osd_op_effects applying watch connect on session "
-	     << session.get() << " watcher " << watcher << dendl;
+	     << (session ? session.get() : nullptr) << " watcher " << watcher << dendl;
```

Comment:
If there is no session here would it make more sense to just call watchers.erase and remove the watcher?

Reply:
the idea is to keep the original code logic. for example, session was gone quickly seems it can also be understood as the session timeout.
in fact, i haven't looked at the logic of osd watcher carefully, so keep the original code logic as much as possible to avoid missing something, such as watcher persistence. :(

Reply:
> If there is no session here would it make more sense to just call watchers.erase and remove the watcher? or, if it's not in watchers yet, just clean up the reference in ctx->watch_connects?

looks like watcher is saved into attr. so if there is no session, we have to rmattr.

Reply:
> so if there is no session, we have to rmattr.

Not rmattr, but update the object info.  The problem, as I understand it, is that the watch is persisted (object info is updated) in `do_osd_ops`, while the timeout is set only in `Watch::connect` which is called from `do_osd_op_effects` here.  The only way to back out the update to the object info is to arrange for `handle_watch_timeout` to get called.